pkg/virt-launcher/virtwrap/converter/converter.go

```diff
@@ -1449,91 +1449,11 @@ func Convert_v1_VirtualMachineInstance_To_api_Domain(vmi *v1.VirtualMachineInsta
+	if err != nil {
 		return err
 	}
-
-	networks := indexNetworksByName(vmi.Spec.Networks)
-
-	for i, iface := range vmi.Spec.Domain.Devices.Interfaces {
-		net, isExist := networks[iface.Name]
-		if !isExist {
-			return fmt.Errorf("failed to find network %s", iface.Name)
-		}
-
-		if iface.SRIOV != nil {
-			continue
-		}
-
-		ifaceType := getInterfaceType(&vmi.Spec.Domain.Devices.Interfaces[i])
-		domainIface := api.Interface{
-			Model: &api.Model{
-				Type: translateModel(c, ifaceType),
-			},
-			Alias: api.NewUserDefinedAlias(iface.Name),
-		}
-
-		// if UseEmulation unset and at least one NIC model is virtio,
-		// /dev/vhost-net must be present as we should have asked for it.
-		var virtioNetMQRequested bool
-		if mq := vmi.Spec.Domain.Devices.NetworkInterfaceMultiQueue; mq != nil {
-			virtioNetMQRequested = *mq
-		}
-		if ifaceType == "virtio" && virtioNetProhibited {
-			return fmt.Errorf("In-kernel virtio-net device emulation '/dev/vhost-net' not present")
-		} else if ifaceType == "virtio" && virtioNetMQRequested {
-			queueCount := uint(CalculateNetworkQueues(vmi))
-			domainIface.Driver = &api.InterfaceDriver{Name: "vhost", Queues: &queueCount}
-		}
-
-		// Add a pciAddress if specified
-		if iface.PciAddress != "" {
-			addr, err := device.NewPciAddressField(iface.PciAddress)
-			if err != nil {
-				return fmt.Errorf("failed to configure interface %s: %v", iface.Name, err)
-			}
-			domainIface.Address = addr
-		}
-
-		if iface.Bridge != nil || iface.Masquerade != nil {
-			// TODO:(ihar) consider abstracting interface type conversion /
-			// detection into drivers
-
-			// use "ethernet" interface type, since we're using pre-configured tap devices
-			// https://libvirt.org/formatdomain.html#elementsNICSEthernet
-			domainIface.Type = "ethernet"
-			if iface.BootOrder != nil {
-				domainIface.BootOrder = &api.BootOrder{Order: *iface.BootOrder}
-			} else {
-				domainIface.Rom = &api.Rom{Enabled: "no"}
-			}
-		} else if iface.Slirp != nil {
-			domainIface.Type = "user"
-
-			// Create network interface
-			initializeQEMUCmdAndQEMUArg(domain)
-
-			// TODO: (seba) Need to change this if multiple interface can be connected to the same network
-			// append the ports from all the interfaces connected to the same network
-			err := createSlirpNetwork(iface, *net, domain)
-			if err != nil {
-				return err
-			}
-		} else if iface.Macvtap != nil {
-			if net.Multus == nil {
-				return fmt.Errorf("macvtap interface %s requires Multus meta-cni", iface.Name)
-			}
-
-			domainIface.Type = "ethernet"
-			if iface.BootOrder != nil {
-				domainIface.BootOrder = &api.BootOrder{Order: *iface.BootOrder}
-			} else {
-				domainIface.Rom = &api.Rom{Enabled: "no"}
-			}
-		}
-		domain.Spec.Devices.Interfaces = append(domain.Spec.Devices.Interfaces, domainIface)
-	}
-
+	domain.Spec.Devices.Interfaces = append(domain.Spec.Devices.Interfaces, domainInterfaces...)
```

Comment:
Maybe we can pass directly `domain` to a `appendDomainInterfaces` function so we can create some basics for a future pipeline with stages with the interface `func (*api.Domain) error` so we can have an slice of stages in a pipeline.

so we will will have at `network.go`

```golang
func appendInterfaces(domain *api.Domain) error {
if err != nil {
		return err
}
domain.Spec.Devices.Interfaces = append(domain.Spec.Devices.Interfaces, domainInterfaces...)
}
```

and at `converter.go`

```golang
err := appendInterfaces(domain)
if err != nil {
    return err
}
```

Reply:
This will create a mutable function, which I usually try to avoid as much as possible.
`createDomainInterfaces` is an immutable function that will not change the state of any of the input.

Mutable functions have an inner side-effect which as a reader cannot be assured what it touched/changed.
With immutable functions, you are assured nothing got changed.

Unfortunately, unlike other languages, with Golang it is less convenient to declare a function or its arguments immutable (except copying the whole data/struct).

But maybe a future refactoring will present a reason that will worth such a change. I just do not see the benefit of it now.